src/riscv/lib/src/jit/builder/instr_map.rs

```diff
@@ -0,0 +1,235 @@
+    }
+}
+
+impl Index<InstrId> for AddrMap {
+    type Output = Address;
+
+    fn index(&self, index: InstrId) -> &Self::Output {
+        &self.addresses[index.0]
+    }
+}
+
+/// An array holding instr-info types with a matching [`AddrMap`].
+///
+/// Paired items across the types are located by the same [`InstrId`].
+#[derive(Debug, Clone)]
+pub struct InstrMap<V> {
+    instructions: Vec<V>,
+}
+
+impl<V> InstrMap<V> {
+    /// Iterate over the instructions, yielding pairs of (`InstrId`, `&V`).
+    pub fn iter(&self) -> impl Iterator<Item = (InstrId, &V)> {
+        self.instructions
+            .iter()
+            .enumerate()
+            .map(|(idx, value)| (InstrId(idx), value))
+    }
+
+    /// Iterate over the instructions, yielding pairs of (`InstrId`, `&mut V`).
+    pub fn iter_mut(&mut self) -> impl Iterator<Item = (InstrId, &mut V)> {
+        self.instructions
+            .iter_mut()
+            .enumerate()
+            .map(|(idx, value)| (InstrId(idx), value))
+    }
+
+    /// Transform the values in the `InstrMap` using the provided function, producing a new `InstrMap`.
+    pub fn map<F, U>(self, f: F) -> InstrMap<U>
+    where
+        F: Fn(InstrId, &V) -> U,
+    {
+        let instructions = self
+            .iter()
+            .map(|(instr_id, val)| f(instr_id, val))
+            .collect();
+        InstrMap { instructions }
+    }
+}
+
+impl<V> Index<InstrId> for InstrMap<V> {
+    type Output = V;
+
+    fn index(&self, index: InstrId) -> &Self::Output {
+        &self.instructions[index.0]
+    }
+}
+
+impl<V> IndexMut<InstrId> for InstrMap<V> {
+    fn index_mut(&mut self, index: InstrId) -> &mut Self::Output {
+        &mut self.instructions[index.0]
+    }
+}
+
+/// Builder for initialising an ([`AddrMap`], [`InstrMap`]) pair.
+/// The builder ensures that the two maps are consistent with each other.
+/// The keys of the instruction BTreeMap are the addresses, and the values are the
+/// associated instruction information.
+pub struct InstrMapBuilder<T> {
+    instructions: BTreeMap<Address, T>,
+}
+
+impl<T: Clone> InstrMapBuilder<T> {
+    /// Create a new builder with the specified capacity.
+    pub fn new() -> Self {
+        Self {
+            instructions: BTreeMap::new(),
+        }
+    }
+
+    /// Insert a new address and instruction pair to the builder. If the address
+    /// already exists, the existing instruction will be replaced with the new one.
+    pub fn insert(&mut self, address: Address, instruction: T) {
+        self.instructions.insert(address, instruction);
+    }
+
+    /// Finalise the builder and produce the (`AddrMap`, `InstrMap`) pair.
+    pub fn build(self) -> (AddrMap, InstrMap<T>) {
+        let addr_map = AddrMap {
+            addresses: self.instructions.keys().cloned().collect(),
+        };
+        let instr_map = InstrMap {
+            instructions: self.instructions.values().cloned().collect(),
+        };
```

Comment:
you can use `self.instructions.into_iter()` which will give you an owning iterator that empties both the keys and the values.  In the hopes that "draining" the map can be more optimised by the compiler/std crate.